engine/src/witness/sol/nonce_witnessing.rs

```diff
@@ -43,41 +48,48 @@ impl<Inner: ChunkedByVault> ChunkedByVaultBuilder<Inner> {
 		state_chain_runtime::Runtime: RuntimeHasChain<Inner::Chain>,
 		state_chain_runtime::RuntimeCall:
 			RuntimeCallHasChain<state_chain_runtime::Runtime, Inner::Chain>,
+		StateChainClient: SignedExtrinsicApi + ChainApi + StorageApi + Send + Sync + 'static,
 	{
-		self.then(move |_epoch, header| {
+		self.then(move |epoch, header| {
 			assert!(<Inner::Chain as Chain>::is_block_witness_root(header.index));
 			let sol_rpc = sol_rpc.clone();
 			let _process_call = process_call.clone();
-			let nonce_accounts = nonce_accounts.clone();
+			let state_chain_client = state_chain_client.clone();
 			async move {
-				let nonce_addresses: Vec<SolAddress> = nonce_accounts
-					.clone()
-					.into_iter()
-					.map(|(nonce_address, _)| nonce_address)
-					.collect();
+				let nonce_accounts: HashMap<SolAddress, SolHash> = state_chain_client
+					.storage_map::<pallet_cf_environment::SolanaUnAvailableNonceAccounts<
```

Comment:
Can we think about any scenario in which also checking the AvailableNonceAccounts could avoid an edge case? Of course that shouldn't happen in normal circumstances, ideally never, but just throwing it out there in case anyone can think about one.